packages/protocol/contracts/L1/ProverPool.sol

```diff
@@ -0,0 +1,479 @@
+    //     re-use his funds currently in the 'staking queue'. So basically he is out of the
+    //     exiting queue.
+    //
+    //     B: Almost same situation as above, except that when Bob is leaving (force leaving)
+    //     Carol decided to leave as well, so she is not in the topProver array anymore. It means
+    //     Bob could come back with his stake (currenlty in the exit 'queue') or even less than that
+    //     amount since Carol left, there is 1 "free space". So he could decide to re-stake less than
+    //     his amount in the current exit 'queue', so technically he still have some left in the exit queue
+    //     while also being a topProver in the same time!
+
+
+    // This is 1/4 slot
+    // TODO: Dani, this structure should hold all information required to
+    // dynamically calculate the weight of the top 32 provers and select one
+    // based on a random number.
+    struct TopProver {
+        uint32 amount; // this value will change when we slash the prover
+        uint16 rewardPerGas;
+        uint16 currentCapacity;
+    }
+
+    // prover always have one Prover object,regardless if it is active or exiting.
+    // @TODO: We do not need to use both. We shall have 1 for TopProvers, and 1 for ExitingProvers
+    struct ExitingProver {
+        uint64 exitRequestedAt;
+        uint32 exitAmount;
+        // @QUESTION: I think 'usedCapacity' is not needed here or at least cannot be calculated only from the
+        // current capacity. Because current capacity always deducted when prover is selected to
+        // prove. We can only monitor the 'usedCapacity' in case we increase it once currentCapacity 
+        // deducted, but that increases writes. So simply we shall allow 'exit' with currentCapacity
+        // Then when (if) they reenter, they need to specify their actually available capacity in the exact moment
+        // they are reentering. If possible, try to avoid extra complexity into the contracts.
+        //uint16 usedCapacity;
+    }
+
+    // TOP PROVERS LIST
+    TopProver[32] public topProvers; // 32/4 = 8 slots
+    // EXITING PROVER MAPPING
+    mapping(address prover => ExitingProver) exitingProvers;
+
+    // Id mappings
+    mapping(uint8 id => address) idToProver;
+    mapping(address prover => uint8 id) proverToId;
+
+    // Exit period. Might fine tune later on !
+    uint256 public constant EXIT_PERIOD = 1 weeks;
+
+    // 500 means 5% if 10_000 is 100 %
+    uint256 public constant SLASH_AMOUNT_IN_BP = 500; 
+
+    uint256 public WEI_TO_1_TKO = 10e8;
+
+    uint256[100] private __gap;
+
+    event ProverEntered(
+        address prover, uint32 amount, uint16 rewardPerGas, uint16 capacity
+    );
+
+    event ProverChangedParameters(address prover, uint32 newBalance, uint16 newReward, uint16 newCapacity);
+
+    event ProverKickeOutByWithAmount(address kickedOut, address newProver, uint32 totalAmount);
```

Comment:
We do not need ProverKickeOutByWithAmount, 
We can use two events, ProverExited + ProverStaked.

I suggest to remove "Prover" prefix, simply use the following events:

Staked/Exiting/Exited/Slashed